### PR TITLE
Make password optional

### LINKED ISSUE #195: Trying to connect to a database without password
I have no password for the root user in my development machine, but I'm unable to run the migrations because of the following error:

```
➜  users-api git:(master) ✗ vapor run migrate
Migrate Command: Prepare
[ ERROR ] MySQL error: Server error: Access denied for user 'root'@'localhost' (using password: YES)
```

I can connect to MySQL server just like always on the terminal using `mysql -uroot`.

This is the code I have in my project:

```swift
app.databases.use(.mysql(
    hostname: Environment.get("DATABASE_HOST") ?? "127.0.0.1",
    username: Environment.get("DATABASE_USERNAME") ?? "root",
    password: Environment.get("DATABASE_PASSWORD") ?? "",
    database: Environment.get("DATABASE_NAME") ?? "vaportest",
    tlsConfiguration: .forClient(certificateVerification: .none)
), as: .mysql)
```

I guess password should be an optional parameter, I tried with `nil` and `.none` but no luck.

I also tried to connect by `url` but I'm unable to find which is the correct format for a valid MySQL url 😟

### PR DESCRIPTION
**Pull request by Lloople**
I'm very new to both Swift and Vapor, but I'm trying my best here so sorry about any possible mistake 🙏.

I tried to add a test for this scenario but I don't know enough yet 😅.

This fixes #195 . I don't have a password on my local environment for the MySQL user and I was unable to run migrations, this is why I think this change should be addressed. 

Thank you for your time and framework.

### CODE REVIEW

**Comment by 0xTim** [comment]
I suspect this will need to be fixed in MySQL-Kit first before being added to Fluent. I think passing an empty string is not the same as passing nothing, but it depends on how MySQL works. I'm going to hook @gwynne in here as she's done a ton of work on MySQL auth recently

**Comment by gwynne** [comment]
Had to test this to be sure, but it turns out that no additional effort at the MySQLKit or MySQLNIO layer is required - in MySQL, "no password" and "the password is an empty string" are the same thing.

I will post a review with further comments.

File: Sources/FluentMySQLDriver/FluentMySQLConfiguration.swift [hunk 1]
```diff
@@ -5,7 +5,7 @@ extension DatabaseConfigurationFactory {
     public static func mysql(
         unixDomainSocketPath: String,
         username: String,
-        password: String,
+        password: String? = nil,
```
Comment (gwynne) [inline Sources/FluentMySQLDriver/FluentMySQLConfiguration.swift:8]:
```suggestion
        password: String?,
```

Do not default the password to nil - we don't want to encourage use of empty passwords.

File: Sources/FluentMySQLDriver/FluentMySQLConfiguration.swift [hunk 2]
```diff
@@ -68,7 +68,7 @@ extension DatabaseConfigurationFactory {
         hostname: String,
         port: Int = 3306,
         username: String,
-        password: String,
+        password: String? = nil,
```
Comment (gwynne) [inline Sources/FluentMySQLDriver/FluentMySQLConfiguration.swift:71]:
```suggestion
        password: String?,
```
Again, don't default to `nil`

**Review by gwynne — CHANGES_REQUESTED** [review]
Unfortunately, making the password parameter an `Optional` is technically API-breaking, as certain usages will not be source-compatible. _However_, this feels like one of those situations similar to the "adding cases to an `enum`" problem where requiring semver-major for something that will not affect 99.9% of users feels excessive, and an exception to make it semver-minor instead seems reasonable. As such, I will defer to @tanner0101's judgement with regards to the semver issue.

More concerning to me is that #195 shows an error being returned with "using password: YES" as part of the message even in the presence of an empty string, even though I have just successfully tested this using an identical configuration locally. It turns out that while this works fine with MySQL 8.0, it does _not_ work with MySQL 5.7. I will have to investigate further.

**Comment by gwynne** [comment]
On further investigation, the difference between 5.7 and 8.0 turns out to be due to 8.0 using `caching_sha2_password` as the default instead of `mysql_native_password`. The bug here is actually at the MySQLNIO layer, which incorrectly encodes a complete auth response for an empty password `""`, when the correct behavior for the `mysql_native_password` plugin is to encode no auth response at all in that case. The existing behavior is already correct for the caching SHA-2 plugin.

As such, this PR is unnecessary; I am closing it out in favor of the one I'm about to open against MySQLNIO. @Lloople Your effort and contributions are very much appreciated, and my closing this PR out is in no way intended to discourage you! Please don't hesitate to open new issues and PRs in the future; they are always welcome- this one definitely served to point me in the right direction to fix the problem!

**Comment by Lloople** [comment]
Thank you very much for your time and for such an explanation about the issue! 😍  I'll try my best to continue helping with anything I can

**Pull request closed without merging**